File: Baekjoon/1303_전쟁 - 전투/_1303_전쟁_전투.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class _1303_전쟁_전투 {
	static int bfs(String[][] arr, int[][] visited, int[] start, String find) {
		int dx[]= {-1,1,0,0}; // 상하좌우
		int dy[]= {0,0,-1,1};
		
		visited[start[0]][start[1]]=1;
		Queue<int[]> queue=new LinkedList<>();
		queue.add(start);
		
		int cnt=1;
		
		while(queue.size()!=0) {
			int temp[]=queue.poll();
			for(int i=0; i<4; i++) {
				int x=temp[0]+dx[i];
				int y=temp[1]+dy[i];
				if(x>=0 && x<arr.length && y>=0 && y<arr[0].length) { // 범위 안
					if(arr[x][y].equals(find) && visited[x][y]==0) {
						visited[x][y]=1;
						queue.add(new int[] {x,y});
						cnt+=1;
					}
				}
			}
		}
		return cnt;
	}

	public static void main(String[] args) throws IOException{
		BufferedReader bf=new BufferedReader(new InputStreamReader(System.in));
		
		String s=bf.readLine();
		int n=Integer.parseInt(s.split(" ")[0]); // 가로
		int m=Integer.parseInt(s.split(" ")[1]); // 세로
		
		String arr[][]=new String[m][n];
		for(int i=0; i<m; i++) {
			s=bf.readLine();
			for(int j=0; j<n; j++) {
				arr[i][j]=s.split("")[j];
			}
		}
		
		int visited_w[][]=new int[m][n]; // 방문 여부
		int visited_b[][]=new int[m][n];
		int cnt_w=0;
		int cnt_b=0;
		
		for(int i=0; i<m; i++) {
			for(int j=0; j<n; j++) {
				if(arr[i][j].equals("W") && visited_w[i][j]==0) {
					int cnt=bfs(arr, visited_w, new int[] {i,j}, "W");
					cnt_w+=(cnt*cnt);
				}
				if(arr[i][j].equals("B") && visited_b[i][j]==0) {
					int cnt=bfs(arr, visited_b, new int[] {i,j}, "B");
					cnt_b+=(cnt*cnt);
				}
			}
		}
		System.out.println(cnt_w +" "+ cnt_b);

	}

}

File: Baekjoon/1303_전쟁 - 전투/_1303_전쟁_전투.py
import sys
from collections import deque

def bfs(arr, visited, start, find):
    dx=[-1,1,0,0] # 상하좌우
    dy=[0,0,-1,1]

    queue=deque([start])
    visited[start[0]][start[1]]=1
    cnt=1

    while queue:
        temp=queue.popleft()
        for i in range(4):
            x=temp[0]+dx[i]
            y=temp[1]+dy[i]
            if 0<=x<len(arr) and 0<=y<len(arr[0]):
                if arr[x][y]==find and visited[x][y]==0:
                    visited[x][y]=1
                    queue.append([x,y])
                    cnt+=1
    return cnt


if __name__=='__main__':
    n,m=map(int,sys.stdin.readline().split()) # 가로, 세로

    arr=[]
    for i in range(m):
        arr.append(list(sys.stdin.readline().strip()))

    visited_w=[[0]*n for i in range(m)]
    visited_b = [[0] * n for i in range(m)]
    w_cnt=0
    b_cnt=0
    for i in range(m):
        for j in range(n):
            if arr[i][j]=='W' and visited_w[i][j]==0:
                cnt=bfs(arr, visited_w, [i,j], 'W')
                w_cnt+=(cnt**2)
            elif arr[i][j]=='B' and visited_b[i][j]==0:
                cnt=bfs(arr, visited_b, [i,j], 'B')
                b_cnt+=(cnt**2)


    print(w_cnt, end=' ')
    print(b_cnt)